commands/applycmd.go

```diff
@@ -35,25 +40,56 @@ func GetApplyRunner(provider provider.Provider, loader manifestreader.ManifestLo
 type ApplyRunnerWrapper struct {
 	applyRunner *apply.ApplyRunner
 	factory     cmdutil.Factory
+	ioStreams   genericclioptions.IOStreams
+	applyCRD    bool // Apply the ResourceGroup CRD before applying
 }
 
 // Command returns the wrapped ApplyRunner cobraCommand structure.
 func (w *ApplyRunnerWrapper) Command() *cobra.Command {
 	return w.applyRunner.Command
 }
 
-// RunE runs the ResourceGroup CRD installation as a pre-step if an
-// environment variable exists. Then the wrapped ApplyRunner is
-// invoked. Returns an error if one happened. Swallows the
-// "AlreadyExists" error for CRD installation.
+// RunE delegates to the stored applyRunner. Before the delegation, this
+// function either applies the inventory ResourceGroup CRD
+// (--apply-inventory-crd flag), or checks if the inventory
+// ResourceGroup CRD is available. Returns an error if one occurs in the
+// delegation.
 func (w *ApplyRunnerWrapper) RunE(cmd *cobra.Command, args []string) error {
-	if _, exists := os.LookupEnv(resourceGroupEnv); exists {
-		klog.V(4).Infoln("wrapper applyRunner detected environment variable")
-		err := live.ApplyResourceGroupCRD(w.factory)
-		if err != nil && !apierrors.IsAlreadyExists(err) {
-			return err
-		}
+	if w.applyCRD {
+		// Force the apply of the inventory ResourceGroup CRD before applying.
+		// This does nothing (no error) if the CRD already exists.
+		w.ApplyResourceGroupCRD()
+	} else if !live.ResourceGroupCRDApplied(w.factory) {
+		// Otherwise, report the inventory ResourceGroup if missing.
+		fmt.Fprintln(w.ioStreams.Out, "inventory ResourceGroup CRD is missing")
+		fmt.Fprintln(w.ioStreams.Out, "run 'kpt live install-resource-group' to remedy")
 	}
-	klog.V(4).Infoln("wrapper applyRunner run...")
 	return w.applyRunner.RunE(cmd, args)
 }
+
+func (w *ApplyRunnerWrapper) ApplyResourceGroupCRD() {
+	ch := live.ApplyResourceGroupCRD(w.factory)
+	// Event channel will close when tasks are completed, error, or timeout occurs.
+	var firstApply = true
+	var succeeded = true
+	for e := range ch {
+		if e.Type == event.ErrorType {
+			err := e.ErrorEvent.Err
+			if apierrors.IsAlreadyExists(err) {
+				firstApply = false
+			} else {
+				succeeded = false
+			}
+		}
+	}
+	// Print success or failure output for ResourceGroup apply. A failure
+	// must NOT stop the following 'applyRunner.RunE'.
+	if firstApply {
+		if succeeded {
+			fmt.Fprintf(w.ioStreams.Out, "applying inventory ResourceGroup CRD...success\n")
```

Comment:
Nit: This looks a bit strange as the output suggests we are printing this before we actually make the change.

Reply:
I don't quite understand this comment.

Reply:
Fixed.